### PR TITLE
test(auto labeling 3d): add test for filter objects

### PR DESCRIPTION
**Pull request by Shin-kyoto**
## Summary

Add test for `filter_objects` to prevent regression and make refactoring easier.

## Change point

- test(auto labeling 3d): add test for filter objects

## Note

This PR influences `auto_labeling_3d` only.

## Test performed

I checked this PR by running the test.

<details>
<summary>test log</summary>

```sh
root@8582982598da:/workspace# python -m unittest discover tools/auto_labeling_3d/filter_objects/test/ -v
test_ensemble_basic_functionality (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_basic_functionality) ... ok
test_ensemble_empty_predictions (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_empty_predictions) ... ok
test_ensemble_label_groups (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_label_groups) ... ok
test_ensemble_partial_empty_predictions (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_partial_empty_predictions) ... ok
test_ensemble_single_model (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_single_model) ... ok
test_ensemble_with_overlapping_predictions (test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_with_overlapping_predictions) ... ok
test_extreme_iou_thresholds (test_nms_ensemble_model.TestNMSEnsembleModel.test_extreme_iou_thresholds) ... ok
test_init (test_nms_ensemble_model.TestNMSEnsembleModel.test_init) ... ok
test_weight_mismatch_assertion (test_nms_ensemble_model.TestNMSEnsembleModel.test_weight_mismatch_assertion) ... ok
test_calculate_iou_identical_boxes (test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_identical_boxes) ... ok
test_calculate_iou_multiple_boxes (test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_multiple_boxes) ... ok
test_calculate_iou_no_overlap (test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_no_overlap) ... ok
test_calculate_iou_partial_overlap (test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_partial_overlap) ... ok
test_nms_indices_basic (test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_basic) ... ok
test_nms_indices_empty_input (test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_empty_input) ... ok
test_nms_indices_no_overlap (test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_no_overlap) ... ok
test_nms_indices_single_box (test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_single_box) ... ok
test_nms_indices_various_thresholds (test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_various_thresholds) ... ok
test_filter_and_weight_instances_no_target_labels (test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_no_target_labels) ... ok
test_filter_and_weight_instances_with_empty_instances (test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_with_empty_instances) ... ok
test_filter_and_weight_instances_with_target_labels (test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_with_target_labels) ... ok
test_basic_filtering (test_threshold_filter.TestThresholdFilter.test_basic_filtering) ... ok
test_boundary_values (test_threshold_filter.TestThresholdFilter.test_boundary_values) ... ok
test_empty_data (test_threshold_filter.TestThresholdFilter.test_empty_data) ... ok
test_extreme_thresholds (test_threshold_filter.TestThresholdFilter.test_extreme_thresholds) ... ok
test_filter_statistics_logging (test_threshold_filter.TestThresholdFilter.test_filter_statistics_logging) ... ok
test_init (test_threshold_filter.TestThresholdFilter.test_init) ... ok
test_init_with_duplicate_use_label (test_threshold_filter.TestThresholdFilter.test_init_with_duplicate_use_label) ... ok
test_multiple_frames (test_threshold_filter.TestThresholdFilter.test_multiple_frames) ... ok
test_should_filter_instance_private_method (test_threshold_filter.TestThresholdFilter.test_should_filter_instance_private_method) ... ok
test_unknown_class_handling (test_threshold_filter.TestThresholdFilter.test_unknown_class_handling) ... ok
test_use_label_filtering (test_threshold_filter.TestThresholdFilter.test_use_label_filtering) ... ok
test_various_thresholds (test_threshold_filter.TestThresholdFilter.test_various_thresholds) ... ok

----------------------------------------------------------------------
Ran 33 tests in 0.098s

OK
```

```sh
root@8582982598da:/workspace# python -m unittest tools.auto_labeling_3d.filter_objects.test.test_threshold_filter -v
test_basic_filtering (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_basic_filtering) ... ok
test_boundary_values (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_boundary_values) ... ok
test_empty_data (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_empty_data) ... ok
test_extreme_thresholds (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_extreme_thresholds) ... ok
test_filter_statistics_logging (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_filter_statistics_logging) ... ok
test_init (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_init) ... ok
test_init_with_duplicate_use_label (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_init_with_duplicate_use_label) ... ok
test_multiple_frames (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_multiple_frames) ... ok
test_should_filter_instance_private_method (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_should_filter_instance_private_method) ... ok
test_unknown_class_handling (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_unknown_class_handling) ... ok
test_use_label_filtering (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_use_label_filtering) ... ok
test_various_thresholds (tools.auto_labeling_3d.filter_objects.test.test_threshold_filter.TestThresholdFilter.test_various_thresholds) ... ok

----------------------------------------------------------------------
Ran 12 tests in 0.016s

OK
```

```sh
root@8582982598da:/workspace# python -m unittest tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model -v
test_ensemble_basic_functionality (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_basic_functionality) ... ok
test_ensemble_empty_predictions (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_empty_predictions) ... ok
test_ensemble_label_groups (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_label_groups) ... ok
test_ensemble_partial_empty_predictions (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_partial_empty_predictions) ... ok
test_ensemble_single_model (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_single_model) ... ok
test_ensemble_with_overlapping_predictions (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_ensemble_with_overlapping_predictions) ... ok
test_extreme_iou_thresholds (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_extreme_iou_thresholds) ... ok
test_init (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_init) ... ok
test_weight_mismatch_assertion (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSEnsembleModel.test_weight_mismatch_assertion) ... ok
test_calculate_iou_identical_boxes (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_identical_boxes) ... ok
test_calculate_iou_multiple_boxes (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_multiple_boxes) ... ok
test_calculate_iou_no_overlap (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_no_overlap) ... ok
test_calculate_iou_partial_overlap (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_calculate_iou_partial_overlap) ... ok
test_nms_indices_basic (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_basic) ... ok
test_nms_indices_empty_input (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_empty_input) ... ok
test_nms_indices_no_overlap (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_no_overlap) ... ok
test_nms_indices_single_box (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_single_box) ... ok
test_nms_indices_various_thresholds (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSHelperFunctions.test_nms_indices_various_thresholds) ... ok
test_filter_and_weight_instances_no_target_labels (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_no_target_labels) ... ok
test_filter_and_weight_instances_with_empty_instances (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_with_empty_instances) ... ok
test_filter_and_weight_instances_with_target_labels (tools.auto_labeling_3d.filter_objects.test.test_nms_ensemble_model.TestNMSModelInstances.test_filter_and_weight_instances_with_target_labels) ... ok

----------------------------------------------------------------------
Ran 21 tests in 0.069s

OK
```

</details>

### CODE REVIEW

File: tools/auto_labeling_3d/docs/filter_objects.md [hunk 1]
```diff
@@ -1,5 +1,116 @@
...
+
+#### filter
+
+- Function containing the main filtering logic
+- Processes the entire dataset and applies filtering to each frame
+- Must return filtered dataset info with statistics
+
+#### _should_filter_instance
+
+- Function that determines whether a single instance should be filtered out
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/docs/filter_objects.md]: Remove `Function` same reason above

File: tools/auto_labeling_3d/docs/filter_objects.md [hunk 2]
```diff
@@ -8,11 +119,103 @@
...
+
+**Important**: Weighted scores are used for NMS selection only. Original confidence scores are preserved in the output instances.
+
+#### NMS Process Cases
+
+##### Case 1: High IoU (Overlap) - Suppression
+
+![NMS Ensemble: Overlap Case](./figs/nms_ensemble_with_overlap.svg)
+
+When two detections have high IoU (> threshold), the detection with lower weighted score is suppressed, but the original confidence score is preserved in the output.
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/docs/filter_objects.md]: When `more than one detections`

File: tools/auto_labeling_3d/docs/filter_objects.md [hunk 3]
```diff
@@ -8,11 +119,103 @@
...
+    L --> M
+```
+
+**Important**: Weighted scores are used for NMS selection only. Original confidence scores are preserved in the output instances.
+
+#### NMS Process Cases
+
+##### Case 1: High IoU (Overlap) - Suppression
+
+![NMS Ensemble: Overlap Case](./figs/nms_ensemble_with_overlap.svg)
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/docs/filter_objects.md]: For NMS, did you implement axis-aligned NMS or Non-axis-aligned NMS

File: tools/auto_labeling_3d/filter_objects/test/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+#### 3. Helper Functions
+- **IoU Calculation**: Bird's Eye View intersection over union
+- **NMS Algorithm**: Non-maximum suppression with various thresholds
+
+## Running Tests
+
+```bash
+# All tests (42 total)
+python -m pytest tools/auto_labeling_3d/filter_objects/test/ -v
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/README.md]: Any reason we use `pytest`? I would recommend the default `unittest.TestCase`

File: tools/auto_labeling_3d/filter_objects/test/conftest.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+import logging
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/conftest.py]: We should keep everything in one unit test class as much as possible, let's move them to their respective unit test file

Comment (KSeangTan) [reply tools/auto_labeling_3d/filter_objects/test/conftest.py]: For example, we should have the following:

```
class TestNMSModelInstances(unitest.TestCase):
   
    def setUp(self):
        """Variables or setup need for every unit test"""
    
    def tearDown(self):
        """Clean up every unit test. """
    
    @classmethod
    def setUpClass(cls):
       """Class-wise set up. Only init once. """
    
    @classmethod
    def tearDownClass(cls):
        """Clean up class-wise. Only init once. """
```

for a cleaner design

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+
+
+class TestNMSModelInstances:
+    """Test cases for NMSModelInstances dataclass."""
+
+    def test_filter_and_weight_instances_basic(self, sample_classes):
+        """Test basic filtering and weighting of instances in NMSModelInstances.
+
+        Verifies: Instance filtering respects class membership and score weighting is applied
+        Expected: Only instances of specified classes are kept with scores adjusted by model weight
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: fullstop

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+
+        model_instances = NMSModelInstances(
+            model_id=0, instances=instances, weight=0.8, class_name_to_id=class_name_to_id
+        )
+
+        target_label_names = ["car", "pedestrian"]
+        filtered_instances, boxes, scores = model_instances.filter_and_weight_instances(target_label_names)
+
+        # Should keep car and pedestrian, filter out bicycle
+        assert len(filtered_instances) == 2
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: Another good reason to use `unittest.TestCase`

We can simply call:
```
self.assertEqual(len(filtered_instances), 2)
```

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+    NMSModelInstances,
+    _calculate_iou,
+    _nms_indices,
+)
+
+
+class TestNMSModelInstances:
+    """Test cases for NMSModelInstances dataclass."""
+
+    def test_filter_and_weight_instances_basic(self, sample_classes):
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: The name `basic` is too broad, and it's not a good unit test, for a good unit test, we need to narrow down as much as possible.

**Review by KSeangTan — CHANGES_REQUESTED** [review]
I didn't read all unit tests yet, but please resolve the comments for this round of reviews first

File: tools/auto_labeling_3d/filter_objects/test/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+## Directory Structure
+
+```
+test/
+├── conftest.py              # Shared fixtures and test data
+├── test_threshold_filter.py # ThresholdFilter tests
+├── test_nms_ensemble_model.py # NMSEnsembleModel tests
+└── README.md               # This file
+```
```
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/README.md]: Thank you for your feedback. I'd like to share a slightly different perspective on test code organization.

I believe that when test codes are scattered across different modules, it can become challenging to locate and maintain them effectively. I think it would be beneficial to keep test code and implementation code separated, which could reduce the effort needed to find specific tests.

For example, PyTorch follows a similar approach where tests are organized in a dedicated directory structure: [https://github.com/pytorch/pytorch/tree/main/test](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)

While some developers might find it convenient to have implementation and tests in the same directory, I think there are likely more users who would find it problematic when test codes are dispersed throughout the codebase.

Of course, I'd be interested in hearing the project's policy and other team members' thoughts on this matter. What do you think?

File: tools/auto_labeling_3d/filter_objects/test/README.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+- **Basic Filtering**: Confidence threshold application (≥ threshold kept)
+- **Boundary Values**: Edge cases at exact threshold values  
+- **Empty Data**: Handling of empty input/output scenarios
+- **Multi-frame**: Processing multiple time frames
+- **Unknown Classes**: Graceful handling of unmapped classes
+- **Statistics Logging**: Verification of filtering statistics
+
+## NMSEnsembleModel Tests
+
+Tests Non-Maximum Suppression ensemble functionality:
```
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/README.md]: Thank you for your comment!
I removed in https://github.com/tier4/AWML/pull/122/commits/cea0ab43865f8ba6b25d3dd96920022cd37dc2d7

File: tools/auto_labeling_3d/docs/filter_objects.md [hunk 11]
```diff
@@ -1,5 +1,116 @@
...
+    +filter()
+    +_should_filter_instance()
+  }
+
+  BaseFilter <|-- ThresholdFilter : inheritance
+```
+
+#### filter
+
+- Function containing the main filtering logic
```
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/docs/filter_objects.md]: Thank you for your comment!
I removed in https://github.com/tier4/AWML/commit/57455ead21ac3b792c624ecd1489b7ab00a154d1

File: tools/auto_labeling_3d/docs/filter_objects.md (hunk 2 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/docs/filter_objects.md]: Thank you for your comment!
I changed in https://github.com/tier4/AWML/commit/57455ead21ac3b792c624ecd1489b7ab00a154d1

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+        """Test basic filtering and weighting of instances in NMSModelInstances.
+
+        Verifies: Instance filtering respects class membership and score weighting is applied
```
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: Thank you for your comment!
I changed in https://github.com/tier4/AWML/pull/122/commits/79738fae38cca7cce88aa3f1d5c48ff215900bbd

File: tools/auto_labeling_3d/filter_objects/test/README.md (hunk 4 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/README.md]: memo

The reasons why we use `unittest` are

- consistency
- more pythonic; `unittest` is a part of the Python Standard Library
  - https://www.jetbrains.com/guide/pytest/links/pytest-v-unittest/

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py (hunk 8 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: I changed in https://github.com/tier4/AWML/pull/122/commits/ab17bee0a2900d9b93b0241f84f813da1b9ebe95

File: tools/auto_labeling_3d/filter_objects/test/conftest.py (hunk 5 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/conftest.py]: Thanks!

I changed in https://github.com/tier4/AWML/commit/cf2ff45e5d420a6afdef8eee7ff396a03aab1396

File: tools/auto_labeling_3d/docs/filter_objects.md (hunk 3 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/docs/filter_objects.md]: Axis-aligned NMS is used in this code.

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+    NMSModelInstances,
+    _calculate_iou,
+    _nms_indices,
+)
+
+
+class TestNMSModelInstances(unittest.TestCase):
+    """Test cases for NMSModelInstances dataclass."""
+
+    def setUp(self):
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: Docstring:

```
Initialize states before each individual test method.
```

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+        self.assertIsInstance(boxes, np.ndarray)
+        self.assertIsInstance(scores, np.ndarray)
+        self.assertEqual(boxes.size, 0)
+        self.assertEqual(scores.size, 0)
+
+
+class TestNMSEnsembleModel(unittest.TestCase):
+    """Test cases for NMSEnsembleModel class."""
+
+    def setUp(self):
```
Comment (KSeangTan) [inline tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: Docstring

**Review by KSeangTan — CHANGES_REQUESTED** [review]
Please resolve unit tests, especially, adding more docstring

File: tools/auto_labeling_3d/filter_objects/test/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/README.md]: Thank you so much for discussion in https://github.com/tier4/AWML/issues/144 !
Following this discussion, I'll proceed with the option B for now (since we are developing AWMLv2.)

File: tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py (hunk 13 above)
Comment (Shin-kyoto) [reply tools/auto_labeling_3d/filter_objects/test/test_nms_ensemble_model.py]: @KSeangTan 

Thank you so much for your review!
I added the docstrings in https://github.com/tier4/AWML/pull/122/commits/32b4b2ba7c08a2f1028898494e21e4f8ffa47774 and https://github.com/tier4/AWML/pull/122/commits/861a4b39fed918ae73eb6c4025d1f57076cb5aac

**Review by KSeangTan — APPROVED** [review]
LGTM
